module/VuFind/src/VuFind/Session/SecureDelegator.php

```diff
@@ -0,0 +1,220 @@
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:session_handlers Wiki
+ */
+namespace VuFind\Session;
+
+use VuFind\Db\Table\PluginManager;
+use Zend\Config\Config;
+use Zend\Crypt\BlockCipher;
+use Zend\Math\Rand;
+
+/**
+ * Secure session delegator
+ *
+ * @category VuFind
+ * @package  Session_Handlers
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:session_handlers Wiki
+ */
+class SecureDelegator implements HandlerInterface
+{
+    /**
+     * The block cipher for en/decrypting session data.
+     *
+     * @var BlockCipher
+     */
+    protected $cipher;
+
+    /**
+     * The wrapped session handler.
+     *
+     * @var HandlerInterface
+     */
+    protected $handler;
+
+    /**
+     * SecureDelegator constructor.
+     *
+     * @param HandlerInterface $handler {@see $handler}
+     */
+    public function __construct(HandlerInterface $handler)
+    {
+        $this->handler = $handler;
+        $this->cipher = BlockCipher::factory('openssl');
+    }
+
+    /**
+     * Enable session writing (default)
+     *
+     * @return void
+     */
+    public function enableWrites()
+    {
+        $this->handler->enableWrites();
+    }
+
+    /**
+     * Disable session writing, i.e. make it read-only
+     *
+     * @return void
+     */
+    public function disableWrites()
+    {
+        $this->handler->disableWrites();
+    }
+
+    /**
+     * Set configuration.
+     *
+     * @param Config $config Session configuration ([Session] section of
+     * config.ini)
+     *
+     * @return void
+     */
+    public function setConfig(Config $config)
+    {
+        $this->handler->setConfig($config);
+    }
+
+    /**
+     * Closes the wrapped handler;
+     *
+     * @return bool
+     */
+    public function close()
+    {
+        return $this->handler->close();
+    }
+
+    /**
+     * Destroys a session.
+     *
+     * @param string $session_id Session id
+     *
+     * @return bool
+     */
+    public function destroy($session_id)
+    {
+        return $this->handler->destroy($session_id);
+    }
+
+    /**
+     * Triggers garbage collection.
+     *
+     * @param int $maxlifetime Session maximum lifetime.
+     *
+     * @return bool
+     */
+    public function gc($maxlifetime)
+    {
+        return $this->handler->gc($maxlifetime);
+    }
+
+    /**
+     * Opens a session.
+     *
+     * @param string $save_path Session save path
+     * @param string $name      Session name
+     *
+     * @return bool
+     */
+    public function open($save_path, $name)
+    {
+        $cookie = "{$name}_KEY";
+
+        if (!isset($_COOKIE[$cookie])) {
```

Comment:
Are you aware of the VuFind\Cookie\CookieManager class/service? It provides a level of abstraction on top of $_COOKIE and handles some config.ini settings around cookie path, etc. It would probably be preferable to use that here instead of the lower-level functionality, unless there is some technical reason that it is not possible.

Reply:
`VuFind\Cookie\CookieManager` service now gets used to get/set the cookie.